File: src/modules/app/api/settings/services/vehicle-category-service.ts
import { CreateVehicleCategory } from '@/modules/app/api/settings/data-transfer-objects/create-vehicle-category';
import { UpdateVehicleCategory } from '@/modules/app/api/settings/data-transfer-objects/update-vehicle-category';
import { http } from '@/core';
import { VehicleCategory } from '@/modules/app/api/settings/entities/vehicle-category';

export class VehicleCategoryService {
    public async getVehicleCategories() {
        try {
            var res = await http.get<[{ id: string; name: string; description?: string }]>(
                '/settings/vehicle-category'
            );

            return res.data.map(vc => {
                var newOne = new VehicleCategory(vc.name, vc.description);
                newOne.id = vc.id;

                return newOne;
            });
        } catch (e) {
            if (e.response.status == 404) {
                return [];
            } else {
                throw e;
            }
        }
    }

    public async createVehicleCategory(create: CreateVehicleCategory) {
        var res = await http.post('/settings/vehicle-category', create);

        var vc = new VehicleCategory(res.data.name, res.data.description);
        vc.id = res.data.id;

        return vc;
    }

    public async updateVehicleCategory(update: UpdateVehicleCategory) {
        var res = await http.put(`/settings/vehicle-category/${update.id}`, update);

        var vc = new VehicleCategory(res.data.name, res.data.description);
        vc.id = res.data.id;

        return vc;
    }

    public async deleteVehicleCategory(vehicleCategory: VehicleCategory) {
        await http.delete(`/settings/vehicle-category/${vehicleCategory.id}`);
    }
}

File: api/src/Domain/Settings/Services/VehicleCategoryService.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;

namespace DetailingArsenal.Domain.Settings {
    public interface IVehicleCategoryService : IService {
        Task<VehicleCategory> GetById(Guid id);
        Task<List<VehicleCategory>> GetByUser(User user);
        Task<VehicleCategory> Create(CreateVehicleCategory create, User user);
        Task Update(VehicleCategory category, UpdateVehicleCategory update);
        Task Delete(VehicleCategory category);
    }

    public class VehicleCategoryService : IVehicleCategoryService {
        private VehicleCategoryNameUniqueSpecification uniqueNameSpec;
        private VehicleCategoryNotInUseSpecification notInUseSpec;
        private IVehicleCategoryRepo repo;

        public VehicleCategoryService(VehicleCategoryNameUniqueSpecification uniqueNameSpec, VehicleCategoryNotInUseSpecification notInUseSpec, IVehicleCategoryRepo repo) {
            this.uniqueNameSpec = uniqueNameSpec;
            this.notInUseSpec = notInUseSpec;
            this.repo = repo;
        }

        public async Task<VehicleCategory> GetById(Guid id) {
            return await repo.FindById(id) ?? throw new EntityNotFoundException();
        }

        public async Task<List<VehicleCategory>> GetByUser(User user) {
            return await repo.FindByUser(user);
        }

        public async Task<VehicleCategory> Create(CreateVehicleCategory create, User user) {
            var cat = VehicleCategory.Create(create, user);

            await uniqueNameSpec.CheckAndThrow(cat);
            await repo.Add(cat);

            return cat;
        }

        public async Task Update(VehicleCategory category, UpdateVehicleCategory update) {
            category.Name = update.Name;
            category.Description = update.Description;

            await uniqueNameSpec.CheckAndThrow(category);
            await repo.Update(category);
        }

        public async Task Delete(VehicleCategory category) {
            await notInUseSpec.CheckAndThrow(category);
            await repo.Delete(category);
        }
    }
}